vm/systemSmartContracts/validator.go

```diff
@@ -896,9 +900,38 @@ func (v *validatorSC) unStake(args *vmcommon.ContractCallInput) vmcommon.ReturnC
 			v.eei.Finish([]byte{failed})
 			continue
 		}
+
+		numSuccess++
 	}
 
-	err = v.saveRegistrationData(args.CallerAddr, registrationData)
+	return numSuccess
+}
+
+// This is the complete unStake - which after enabling economics V2 will create unStakedFunds on the registration data
+func (v *validatorSC) unStake(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	registrationData, returnCode := v.basicChecksForUnStakeNodes(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	numSuccess := v.unStakeNodesFromStakingSC(args.Arguments, registrationData)
+	if !v.flagEnableTopUp.IsSet() {
+		return vmcommon.Ok
+	}
+
+	// continue by unstaking tokens as well
+	validatorConfig := v.getConfig(v.eei.BlockChainHook().CurrentEpoch())
+	unStakeFromNodes := big.NewInt(0).Mul(validatorConfig.NodePrice, big.NewInt(0).SetUint64(numSuccess))
+	if unStakeFromNodes.Cmp(registrationData.TotalStakeValue) > 0 {
```

Comment:
shouldn't this be an error? In what case can we have unStakeFromNodes > TotalStakeValue ? (changed NodePrice ?)

Reply:
UnStakeTokens before and unStake after.

Reply:
Is the totalStakeValue changed by slashing?

Reply:
yes. slashing has to be moved to the validatorSC. But as there is not yet used, there is no need for it.

Reply:
ok